drivers/i2s/i2s_nrfx.c

```diff
@@ -0,0 +1,1309 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @brief Driver for Nordic Semiconductor nRF I2S
+ */
+
+
+#include <string.h>
+#include <stdlib.h>
+#include <dma.h>
+#include <i2s.h>
+#include <soc.h>
+#include <nrfx.h>
+#include <nrfx_i2s.h>
+#include "i2s_nrfx.h"
+#define LOG_DOMAIN "i2s_nrfx"
+#include <logging/log.h>
+#define LOG_LEVEL _LOG_LEVEL_DBG
+
+LOG_MODULE_REGISTER(i2s_nrfx);
+
+#define LOG_ERROR	LOG_ERR("\r\n[%s] : %u\r\n", __func__, __LINE__)
+#define RING_BUF_INC(idx, limit) {idx = (++idx < limit) ? idx : 0; }
+#define DEV_CFG(dev) \
+	(const struct i2s_nrfx_config *const)((dev)->config->config_info)
+#define DEV_DATA(dev) \
+	((struct i2s_nrfx_data *const)(dev)->driver_data)
+
+enum i2s_if_state {
+	I2S_IF_NOT_READY = 0,
+	I2S_IF_READY,
+	I2S_IF_RESTARTING,
+	I2S_IF_RUNNING,
+	I2S_IF_STOPPING,
+	I2S_IF_NEEDS_RESTART,
+	I2S_IF_ERROR
+};
+
+struct i2s_nrfx_config {
+	u8_t sck_pin;
+	u8_t lrck_pin;
+	u8_t mck_pin;
+	u8_t sdout_pin;
+	u8_t sdin_pin;
+	void (*instance_init)(struct device *dev);
+};
+
+struct queue_item {
+	void *data;
+	u32_t size;
+};
+
+struct queue {
+	struct queue_item *queue_items;
+	u8_t read_idx;
+	u8_t write_idx;
+	u8_t len;
+};
+
+struct channel_str {
+	struct k_sem sem;
+	struct k_mem_slab *mem_slab;
+	s32_t timeout;
+	enum i2s_state current_state;
+	struct queue mem_block_queue;
+	enum i2s_trigger_cmd last_trigger_cmd;
+	int (*start)(void);
```

Comment:
Why the function pointers are stored here (in RAM)? Are they supposed to be modified?